File: 295_Find_Median_from_Data_Stream.go
// 2020/4/18
package main

import (
	"fmt"
	// "math"
	"container/heap"
	"time"
)

func main() {
	start := time.Now()
	m := MedianFinder{}
	m.AddNum(1)
	m.AddNum(2)
	fmt.Println(m.FindMedian() == 1.5)
	m.AddNum(3)
	fmt.Println(m.FindMedian() == 2)
	elapsed := time.Now().Sub(start)
	fmt.Println(elapsed)
}

type minHeap []int

func (h minHeap) Len() int           { return len(h) }
func (h minHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h minHeap) Less(i, j int) bool { return h[i] < h[j] }

func (h *minHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *minHeap) Pop() interface{} {
	var x interface{}
	x, *h = (*h)[len(*h)-1], (*h)[:len(*h)-1]
	return x
}

type MedianFinder struct {
	even bool
	minh *minHeap
	maxh *minHeap
}

/** initialize your data structure here. */
func Constructor() MedianFinder {
	minh := &minHeap{}
	heap.Init(minh)
	maxh := &minHeap{}
	heap.Init(maxh)
	return MedianFinder{
		minh: minh,
		maxh: maxh,
	}
}

func (mf *MedianFinder) AddNum(num int) {
	fmt.Println(num)
	heap.Push(mf.minh, num)
	heap.Push(mf.maxh, -heap.Remove(mf.minh, 0).(int))

	if mf.minh.Len() < mf.maxh.Len() {
		heap.Push(mf.minh, -heap.Remove(mf.maxh, 0).(int))
	}
}

func (mf *MedianFinder) FindMedian() float64 {
	var median float64
	if mf.minh.Len() > mf.maxh.Len() {
		median = float64((*mf.minh)[0])
	} else {
		median = float64((*mf.minh)[0]-((*mf.maxh)[0])) / 2
	}
	return median
}

File: 295_Find_Median_from_Data_Stream.py
from heapq import heappush, heappop, nsmallest
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.q1, self.q2 = [], []
        
    def get_median(self):
        v1 = -nsmallest(1, self.q1)[0] if self.q1 else None
        v2 =  nsmallest(1, self.q2)[0] if self.q2 else None
        print(v1, v2)
        return v1, v2
    
    def addNum(self, num: int) -> None:
        v1, v2 = self.get_median()
        if len(self.q1) == len(self.q2): 
            if v2 and num > v2: heappop(self.q2); heappush(self.q2, num); heappush(self.q1, -v2)
            else              : heappush(self.q1, -num)
        else:
            if num >= v1: heappush(self.q2, num)
            else: heappop(self.q1); heappush(self.q1, -num); heappush(self.q2, v1); 

    def findMedian(self) -> float:
        v1, v2 = self.get_median()
        if v1 and v2 and len(self.q1) == len(self.q2): return (v1+v2)/2
        elif v1: return v1
        return 0

sol = MedianFinder()
sol.addNum(1)
sol.addNum(2)
print(sol.findMedian()==1.5)
sol.addNum(3)
print(sol.findMedian()==2)
